pkg/block/indexheader/reader_pool.go

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package indexheader
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/tsdb/index"
+	"go.uber.org/atomic"
+
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+type ReaderPool struct {
+	lazyReaderEnabled     bool
+	lazyReaderIdleTimeout time.Duration
+	lazyReaderMetrics     *LazyBinaryReaderMetrics
+	logger                log.Logger
```

Comment:
Wonder if it wouldn't be easier to track open / close metrics here instead of lazy BinaryReader? :thinking: Also we might be missing some gauge for currently loaded headers?

Reply:
> Wonder if it wouldn't be easier to track open / close metrics here instead of lazy BinaryReader?

That was my initial choice, but unfortunately it's more complicated because the pool doesn't have visibility when the lazy load (`open()`) actually occur. Tracking it in `LazyBinaryReader` turned out to be easier and safer.

Reply:
> Also we might be missing some gauge for currently loaded headers?

Technically, it's `indexheader_lazy_load_total - indexheader_lazy_unload_total`. Do we also need a gauge?

Reply:
> Technically, it's indexheader_lazy_load_total - indexheader_lazy_unload_total. Do we also need a gauge?

As long as you subtract metrics from exactly same process, otherwise: `rate(indexheader_lazy_load_total[xx]) - rate(indexheader_lazy_unload_total[xx]` which is getting funky

Anyway, good for now :+1: